activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb

```diff
@@ -629,8 +629,13 @@ def remove_columns(table_name, *column_names, type: nil, **options)
           raise ArgumentError.new("You must specify at least one column name. Example: remove_columns(:people, :first_name)")
         end
 
-        column_names.each do |column_name|
-          remove_column(table_name, column_name, type, **options)
+        if supports_bulk_alter?
```

Comment:
This condition excludes the SQLite adapter, but for the wrong reason: `supports_bulk_alter?` tells us whether the adapter knows how to apply multiple schema changes together when `alter_table` is called with the `bulk: true` option, and not whether the database has the ability to execute an ALTER statement that contains multiple changes.

It's a subtle distinction, but if bulk alter support was added to the SQLite adapter in future (e.g. via https://github.com/rails/rails/pull/31628), this condition would become true, and we'd try to execute a statement that SQLite still couldn't handle. Conversely, if a future version of SQLite added support for dropping multiple columns in a single statement, we couldn't change the adapter to return true from `supports_bulk_alter?`.

I think we should change this method to always use a single statement, and override it in the SQLite adapter to remove each column separately, which is how we deal with its partial ALTER support for other operations: https://github.com/rails/rails/blob/0184a019353fa862c37ae61308e62e4e702c347f/activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb#L240-L283

Reply:
@eugeneius Thank you for that detailed explanation. That makes sense.

I've applied your suggestion in the latest revision. 

Because this adds logic to the SQLite3 adapter, I've restored the query count assert in the test. I think it was helpful in showing me that the override was doing what I expected it to.